util/mcf_dict_util.py

```diff
@@ -0,0 +1,537 @@
+    for cur_file in matching_files:
+        # set of files
+        print(cur_file)
+        if cur_file not in existing_dict or reopen:
+            existing_dict[cur_file] = mcf_file_to_dict_list(cur_file)
+
+    return existing_dict
+
+
+def dict_list_to_mcf_str(node_list: list,
+                         sort_keys: bool = False,
+                         regen_complex_vals: bool = False) -> str:
+    """Creates a string (which could be written to a file) from list of Dict like object representation of nodes.
+
+    Args:
+        node_list: List of OrderedDict or Dict objects representing a node in MCF file.
+        sort_keys: Boolean value to sort the properties before appending the node to the string.
+                    NOTE: This keeps Node on the top and all inline comments would be moved around.
+        regen_complex_vals: Boolean value to regenrate the value string of the property from the list 
+                                of complex value representation.
+        
+    Returns:
+        String representation of the list of nodes.
+    """
+    ret_str = ''
+    for _node in node_list:
+        cur_node = _node.copy()
+
+        is_comment_block = True
+        for prop in cur_node:
+            if not prop.startswith('__comment'):
+                is_comment_block = False
+        # TODO other validations if required
+        if not is_comment_block:
+            if 'Node' not in cur_node:
+                raise ValueError('Each node must have Node: <name>".')
+
+        # preserve comments
+        if '__comment' in cur_node:
+            ret_str += cur_node['__comment']
+        cur_node.pop('__comment', None)
+        # add comments till node
+        prop_list = list(cur_node.keys())
+        i = 0
+        while i < len(prop_list) and prop_list[i].startswith('__comment'):
+            ret_str += cur_node[prop_list[i]]
+            ret_str += '\n'
+            cur_node.pop(prop_list[i], None)
+            i += 1
+
+        if not is_comment_block:
+            # Keep Node: first
+            ret_str += f"{'Node'}: {cur_node['Node']['namespace']+':' if cur_node['Node']['namespace'] else ''}{cur_node['Node']['value']}"
+            ret_str += '\n'
+            cur_node.pop('Node', None)
+
+        prop_list = list(cur_node.keys())
+        # sort keys if flag raised
+        if sort_keys:
+            prop_list = sorted(prop_list)
+        for prop in prop_list:
+            if prop.startswith('__comment'):
+                ret_str += cur_node[prop]
+            else:
+                if 'complexValue' in cur_node[prop] and regen_complex_vals:
+                    cur_node[prop][
+                        'value'] = f"[{' '.join(cur_node[prop]['complexValue'])}]"
+                ret_str += f"{prop}: {cur_node[prop]['namespace']+':' if cur_node[prop]['namespace'] else ''}{cur_node[prop]['value']}"
+            ret_str += '\n'
+        ret_str += '\n'
+    # return ''.join(ret_str.rsplit('\n', 1))
```

Comment:
delete?

Reply:
kept it for future reference in case we want single newline at the end of the file.

Reply:
Then pls add a comment instead of a code block that makes it easier to follow.